src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+
+        return self.validate_user(code, request_handler)
+
+    def is_active(self, user, request_handler):
+        access_token = request_handler.get_secure_cookie('token')
+        if access_token is None:
+            return False
+        access_token = access_token.decode("utf-8")
+
+        if self.user_states.get(user) is None:
+            LOGGER.debug("User %s not found in state" % user)
+            return False
+
+        if self.user_states[user]['state'] is None or self.user_states[user]['state'] != "active":
+            LOGGER.info("User %s state inactive: " % user + str(self.user_states[user]))
+            del self.user_states[user]
+            self.dump_sessions_to_file()
+            return False
+
+        now = time.time()
+        # check session ttl
+        if self.session_expire and (self.user_states[user]['visit'] + self.session_expire) < now:
+            del self.user_states[user]
+            LOGGER.info("User %s session expired, logged out" % user)
+            self.dump_sessions_to_file()
+            return False
+
+        self.user_states[user]['visit'] = now
+
+        # check gitlab response ttl, also check for stale updating (ttl*2)
+        if self.gitlab_update is not None:
+            stale = (self.user_states[user]['updated'] + max(self.gitlab_update*2, 60)) < now
+            ttl_expired = (self.user_states[user]['updated'] + self.gitlab_update) < now
+            updating_now = self.user_states[user]['updating'] is True
+            if ttl_expired and (not updating_now or stale):
+                if self.gitlab_group_support:
+                    self.do_update_groups(user, access_token)
+                else:
+                    self.do_update_user(user, access_token)
+
+        return True
+
+    def get_groups(self, user, known_groups=None):
+        if self.user_states.get(user) is None:
+            return []
+        if self.user_states[user]['groups'] is None:
+            return []
+        return self.user_states[user]['groups']
+
+    def logout(self, user, request_handler):
+        request_handler.clear_cookie('token')
+
+    def clean_expired_sessions(self):
+        now = time.time()
+        if self.session_expire:
+            for userData in list(self.user_states.keys()):
+                if (self.user_states[userData]['visit'] + self.session_expire) < now:
+                    LOGGER.debug("User %s session expired and removed" % userData)
+                    del self.user_states[userData]
+
+    def dump_sessions_to_file(self):
+        if self.gitlab_dump:
+            dumpFile = open(self.gitlab_dump, "w")
+            dumpFile.write(escape.json_encode(self.user_states))
+            dumpFile.close()
+            LOGGER.debug("Dumped state to file %s" % self.gitlab_dump)
+
+    def do_update_user(self, user, access_token):
+        self.user_states[user]['updating'] = True
+        tornado.ioloop.IOLoop.current().spawn_callback(self.update_user_state, user, access_token)
```

Comment:
Out of curiosity: why are you doing it in a separate callback?

Reply:
Pros: User action not blocked while updating user info because it can take time, especially for cloud Gitlab
Cons: If gitlab access revoked user can do some number of requests before logged out
But I decided that pros is more important - lags of each auth_info_ttl seconds is not good